git_repo/services/ext/gogs.py

```diff
@@ -0,0 +1,244 @@
+    fqdn = 'try.gogs.io'
+
+    @classmethod
+    def _url_parse(cls, url):
+        if '://' not in url:
+            url = 'https://'+url
+        parse = urlparse(url)
+        url_base = urlunparse((parse.scheme, parse.netloc)+('',)*4)
+        fqdn = parse.hostname
+        return url_base, fqdn
+
+    @property
+    def url_ro(self):
+        return self.url_base
+
+    @property
+    def url_rw(self):
+        url = self.ssh_url
+        if '@' in url:
+            return url
+        return '@'.join([self.git_user, url])
+
+    def url_api(self, rest):
+        return '{}/api/v1/{}'.format(self.url_base, rest)
+
+    @classmethod
+    def get_auth_token(cls, login, password, prompt=None):
+        import platform
+        name = 'git-repo2 token used on {}'.format(platform.node())
+        if '/' in login:
+            url, login = login.rsplit('/', 1)
+        else:
+            url = input('URL [{}]> '.format(cls.fqdn)) or cls.fqdn
+        url_base, fqdn = cls._url_parse(url)
+        url_api = functools.partial('{}/api/v1/{}'.format, url_base)
+        r = requests.get(url_api('users/{}/tokens'.format(login)), auth=(login, password), verify=False)
+        r.raise_for_status()
+        tokens = r.json()
+        tokens = dict((o['name'], o['sha1']) for o in tokens)
+        if name in tokens:
+            return tokens[name]
+        if 'git-repo2 token' in tokens:
+            return tokens['git-repo2 token']
+        r = requests.post(url_api('users/{}/tokens'.format(login)), auth=(login, password), json=dict(name=name), verify=False)
+        r.raise_for_status()
+        token = r.json()
+        return token['sha1']
+
+    @property
+    def user(self):
+        r = self.session.get(self.url_api('user'))
+        r.raise_for_status()
+        user = r.json()
+        return user['username']
+
+    def orgs(self):
+        r  = self.session.get(self.url_api('user/orgs'))
+        r.raise_for_status()
+        return [o['username'] for o in r.json()]
+
+    def connect(self):
+        self.url_base, self.fqdn = self._url_parse(self.fqdn)
+        verify = self.fqdn == 'try.gogs.io'
+        #verify = True
+        verify = self.config.get('verify', ['no','yes'][verify])
+        if verify.lower().strip() in ('0','no','false',''):
+            verify = False
+        elif verify.lower().strip() in ('1','yes','true'):
+            verify = True
+        self.default_private = self.config.get('default-private', 'true').lower() not in ('0','no','false')
+        self.ssh_url = self.config.get('ssh-url', None) or self.fqdn
+        if not self.repository:
+            config = git_config.GitConfigParser(os.path.join(os.environ['HOME'], '.gitconfig'), True)
+        else:
+            config = self.repository.config_reader()
+        proxies = {}
+        for scheme in 'http https'.split():
+            proxy = config.get_value(scheme, 'proxy', '')
+            if proxy:
+                proxies[scheme] = proxy
```

Comment:
proxies support is a great idea I totally did not think of…

but again, it should be implemented outside of the service implementation, so that it's accessible to all services (the same way `self.insecure` is implemented).

so either you remove it, and make an issue/PR to add this globally, or you leave it, and we'll replace that when we'll implement proxy support across services. Not sure what's the smartest.